diff --git a/pilot/pkg/bootstrap/server.go b/pilot/pkg/bootstrap/server.go
--- a/pilot/pkg/bootstrap/server.go	(revision 8f9263961b52f11f9bfb731e0eeae096bb0acfd0)
+++ b/pilot/pkg/bootstrap/server.go	(date 1636568600190)
@@ -29,6 +29,9 @@
 	"sync"
 	"time"

+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/go-spiffe/v2/workloadapi"
+
 	prometheus "github.com/grpc-ecosystem/go-grpc-prometheus"
 	prom "github.com/prometheus/client_golang/prometheus"
 	"golang.org/x/net/http2"
@@ -180,6 +183,9 @@
 	statusReporter *status.Reporter
 	// RWConfigStore is the configstore which allows updates, particularly for status.
 	RWConfigStore model.ConfigStoreCache
+
+	// source of X.509 certs and bundle, when using SPIFFE Workload API as cert provider
+	x509Source *workloadapi.X509Source
 }

 // NewServer creates a new Server instance based on the provided arguments.
@@ -212,6 +218,16 @@
 	for _, fn := range initFuncs {
 		fn(s)
 	}
+
+	if strings.EqualFold(features.PilotCertProvider, constants.CertProviderSpiffe) {
+		ctx := context.Background()
+		x509Source, err := workloadapi.NewX509Source(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed creating spiffe X.509 source: %v", err)
+		}
+		s.x509Source = x509Source
+	}
+
 	// Initialize workload Trust Bundle before XDS Server
 	e.TrustBundle = s.workloadTrustBundle
 	s.XDSServer = xds.NewDiscoveryServer(e, args.Plugins, args.PodName, args.Namespace)
@@ -931,6 +947,7 @@
 func (s *Server) initIstiodCerts(args *PilotArgs, host string) error {
 	// Skip all certificates
 	var err error
+
 	if hasCustomTLSCerts(args.ServerOptions.TLSOptions) {
 		// Use the DNS certificate provided via args.
 		err = s.initCertificateWatches(args.ServerOptions.TLSOptions)
@@ -954,14 +971,29 @@
 		if err == nil {
 			err = s.initIstiodCertLoader()
 		}
+	} else if strings.EqualFold(features.PilotCertProvider, constants.CertProviderSpiffe) {
+		chain, key, err := s.getCertAndKeyBytes()
+		if err != nil {
+			return err
+		}
+
+		bundleBytes, err := s.getBundleBytes()
+		if err != nil {
+			return err
+		}
+
+		s.setIstioCertBundleAndNotify(chain, key, bundleBytes)
+		s.watchIstioCertUpdates()
+		return nil
 	}
+

 	return err
 }

 // createPeerCertVerifier creates a SPIFFE certificate verifier with the current istiod configuration.
 func (s *Server) createPeerCertVerifier(tlsOptions TLSOptions) (*spiffe.PeerCertVerifier, error) {
-	if tlsOptions.CaCertFile == "" && s.CA == nil && features.SpiffeBundleEndpoints == "" && !s.isDisableCa() {
+	if tlsOptions.CaCertFile == "" && s.CA == nil && features.SpiffeBundleEndpoints == "" && !s.isDisableCa() && !strings.EqualFold(features.PilotCertProvider, constants.CertProviderSpiffe) {
 		// Running locally without configured certs - no TLS mode
 		return nil, nil
 	}
@@ -997,6 +1029,18 @@
 		peerCertVerifier.AddMappings(certMap)
 	}

+	if strings.EqualFold(features.PilotCertProvider, constants.CertProviderSpiffe) {
+		bundleBytes, err := s.getBundleBytes()
+		if err != nil {
+			return nil, err
+		}
+
+		err = peerCertVerifier.AddMappingFromPEM(spiffe.GetTrustDomain(), bundleBytes)
+		if err != nil {
+			return nil, fmt.Errorf("add root CAs into peerCertVerifier failed: %v", err)
+		}
+	}
+
 	return peerCertVerifier, nil
 }

@@ -1063,6 +1107,60 @@
 	return nil
 }

+func (s *Server) watchIstioCertUpdates() {
+	go func() {
+		updatedChan := s.x509Source.Updated()
+		for {
+			<-updatedChan
+			chain, key, err := s.getCertAndKeyBytes()
+			if err != nil {
+				log.Errorf("error watching SPIFFE updates: %v", err)
+				continue
+			}
+
+			bundleBytes, err := s.getBundleBytes()
+			if err != nil {
+				log.Errorf("error watching SPIFFE updates: %v", err)
+				continue
+			}
+
+			s.setIstioCertBundleAndNotify(chain, key, bundleBytes)
+		}
+	}()
+}
+
+func (s *Server) getCertAndKeyBytes() ([]byte, []byte, error) {
+	svid, err := s.x509Source.GetX509SVID()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed fetching X.509 SVID: %v", err)
+	}
+
+	chain, key, err := svid.Marshal()
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to marshal X.509 SVID: %v", err)
+	}
+
+	return chain, key, nil
+}
+
+func (s *Server) getBundleBytes() ([]byte, error) {
+	trustDomain, err := spiffeid.TrustDomainFromString(spiffe.GetTrustDomain())
+	if err != nil {
+		return nil, fmt.Errorf("error trying to parse trust domain %q reason: %v", spiffe.GetTrustDomain(), err)
+	}
+
+	bundle, err := s.x509Source.GetX509BundleForTrustDomain(trustDomain)
+	if err != nil {
+		return nil, fmt.Errorf("unable to find X.509 bundle for trust domain %q: %v", trustDomain, err)
+	}
+
+	bundleBytes, err := bundle.Marshal()
+	if err != nil {
+		return nil, fmt.Errorf("unable to marshal X.509 byndle: %v", err)
+	}
+	return bundleBytes, nil
+}
+
 // StartCA starts the CA or RA server if configured.
 func (s *Server) startCA(caOpts *caOptions) {
 	if s.CA == nil && s.RA == nil {
diff --git a/pkg/istio-agent/xds_proxy.go b/pkg/istio-agent/xds_proxy.go
--- a/pkg/istio-agent/xds_proxy.go	(revision 8f9263961b52f11f9bfb731e0eeae096bb0acfd0)
+++ b/pkg/istio-agent/xds_proxy.go	(date 1636568553844)
@@ -22,6 +22,7 @@
 	"encoding/json"
 	"fmt"
 	"io/ioutil"
+	"istio.io/istio/pkg/security"
 	"math"
 	"net"
 	"net/http"
@@ -701,6 +702,21 @@
 	var certPool *x509.CertPool
 	var rootCert []byte

+	if strings.EqualFold(agent.secOpts.CAProviderName, constants.CertProviderSpiffe) {
+		secretItem, err := agent.secretCache.GenerateSecret(security.RootCertReqResourceName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create TLS dial option with root certificates: %v", err)
+		}
+
+		certPool = x509.NewCertPool()
+		ok := certPool.AppendCertsFromPEM(secretItem.RootCert)
+		if !ok {
+			return nil, fmt.Errorf("failed to create TLS dial option with root certificates")
+		}
+
+		return certPool, nil
+	}
+
 	xdsCACertPath, err := agent.FindRootCAForXDS()
 	if err != nil {
 		return nil, fmt.Errorf("failed to find root CA cert for XDS: %v", err)
diff --git a/pkg/istio-agent/agent.go b/pkg/istio-agent/agent.go
--- a/pkg/istio-agent/agent.go	(revision 8f9263961b52f11f9bfb731e0eeae096bb0acfd0)
+++ b/pkg/istio-agent/agent.go	(date 1636568553837)
@@ -110,7 +110,7 @@
 	envoyWaitCh chan error

 	sdsServer   *sds.Server
-	secretCache *cache.SecretManagerClient
+	secretCache security.SecretProvider

 	// Used when proxying envoy xds via istio-agent is enabled.
 	xdsProxy *XdsProxy
@@ -545,6 +545,9 @@
 	if a.cfg.XDSRootCerts == security.SystemRootCerts {
 		// Special case input for root cert configuration to use system root certificates
 		return "", nil
+	} else if strings.EqualFold(a.secOpts.CAProviderName, constants.CertProviderSpiffe) {
+		// The root cert is provided by the SPIFFE secret manager
+		return "", nil
 	} else if a.cfg.XDSRootCerts != "" {
 		// Using specific platform certs or custom roots
 		rootCAPath = a.cfg.XDSRootCerts
@@ -625,7 +628,12 @@
 }

 // newSecretManager creates the SecretManager for workload secrets
-func (a *Agent) newSecretManager() (*cache.SecretManagerClient, error) {
+func (a *Agent) newSecretManager() (security.SecretProvider, error) {
+	if strings.EqualFold(a.secOpts.CAProviderName, constants.CertProviderSpiffe) {
+		log.Info("Using SPIFFE identity plane")
+		return cache.NewSpiffeSecretManager(a.secOpts)
+	}
+
 	// If proxy is using file mounted certs, we do not have to connect to CA.
 	if a.secOpts.FileMountedCerts {
 		log.Info("Workload is using file mounted certificates. Skipping connecting to CA")
diff --git a/pkg/config/constants/constants.go b/pkg/config/constants/constants.go
--- a/pkg/config/constants/constants.go	(revision 8f9263961b52f11f9bfb731e0eeae096bb0acfd0)
+++ b/pkg/config/constants/constants.go	(date 1636568293867)
@@ -125,4 +125,6 @@
 	// CertProviderNone does not create any certificates for the control plane. It is assumed that some external
 	// load balancer, such as an Istio Gateway, is terminating the TLS.
 	CertProviderNone = "none"
+    // SpiffeCertProvider uses the SPIFFE Workload API to fetch certificates
+	CertProviderSpiffe  = "spiffe"
 )
diff --git a/go.mod b/go.mod
--- a/go.mod	(revision 8f9263961b52f11f9bfb731e0eeae096bb0acfd0)
+++ b/go.mod	(date 1636568293868)
@@ -70,6 +70,7 @@
 	github.com/spf13/cobra v1.2.1
 	github.com/spf13/pflag v1.0.5
 	github.com/spf13/viper v1.8.1
+	github.com/spiffe/go-spiffe/v2 v2.0.0-beta.10
 	github.com/stretchr/testify v1.7.0
 	github.com/xeipuuv/gojsonpointer v0.0.0-20190905194746-02993c407bfb // indirect
 	github.com/yl2chen/cidranger v1.0.2
diff --git a/pkg/security/security.go b/pkg/security/security.go
--- a/pkg/security/security.go	(revision 8f9263961b52f11f9bfb731e0eeae096bb0acfd0)
+++ b/pkg/security/security.go	(date 1636568293868)
@@ -21,6 +21,7 @@
 	"strings"
 	"time"

+	"github.com/spiffe/go-spiffe/v2/bundle/x509bundle"
 	"google.golang.org/grpc/metadata"

 	"istio.io/pkg/env"
@@ -245,6 +246,13 @@
 	GenerateSecret(resourceName string) (*SecretItem, error)
 }

+type SecretProvider interface {
+	SecretManager
+	Close()
+	SetUpdateCallback(func(string))
+	UpdateConfigTrustBundle([]byte) error
+}
+
 // TokenExchanger provides common interfaces so that authentication providers could choose to implement their specific logic.
 type TokenExchanger interface {
 	// ExchangeToken provides a common interface to exchange an existing token for a new one.
@@ -258,6 +266,9 @@

 	RootCert []byte

+	// Trust bundles keyed by trust domain
+	TrustBundles *x509bundle.Set
+
 	// ResourceName passed from envoy SDS discovery request.
 	// "ROOTCA" for root cert request, "default" for key/cert request.
 	ResourceName string
diff --git a/security/pkg/nodeagent/sds/sdsservice.go b/security/pkg/nodeagent/sds/sdsservice.go
--- a/security/pkg/nodeagent/sds/sdsservice.go	(revision 8f9263961b52f11f9bfb731e0eeae096bb0acfd0)
+++ b/security/pkg/nodeagent/sds/sdsservice.go	(date 1636568293868)
@@ -18,6 +18,7 @@
 import (
 	"context"
 	"fmt"
+	"google.golang.org/protobuf/types/known/anypb"
 	"time"

 	"github.com/cenkalti/backoff"
@@ -144,7 +145,11 @@
 			return nil, fmt.Errorf("failed to generate secret for %v: %v", resourceName, err)
 		}

-		res := util.MessageToAny(toEnvoySecret(secret))
+		envoySecret, err := toEnvoySecret(secret)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate secret for %v: %v", resourceName, err)
+		}
+		res := util.MessageToAny(envoySecret)
 		resources = append(resources, &discovery.Resource{
 			Name:     resourceName,
 			Resource: res,
@@ -199,13 +204,22 @@
 }

 // toEnvoySecret converts a security.SecretItem to an Envoy tls.Secret
-func toEnvoySecret(s *security.SecretItem) *tls.Secret {
+func toEnvoySecret(s *security.SecretItem) (*tls.Secret, error) {
 	secret := &tls.Secret{
 		Name: s.ResourceName,
 	}

 	cfg, ok := model.SdsCertificateConfigFromResourceName(s.ResourceName)
 	if s.ResourceName == security.RootCertReqResourceName || (ok && cfg.IsRootCertificate()) {
+		// are there federated bundles?
+		if s.TrustBundles != nil && s.TrustBundles.Len() > 1 {
+			validatorConfig, err := buildSPIFFECertValidatorConfig(s)
+			if err != nil {
+				return nil, err
+			}
+			secret.Type = validatorConfig
+			return secret, nil
+		}
 		secret.Type = &tls.Secret_ValidationContext{
 			ValidationContext: &tls.CertificateValidationContext{
 				TrustedCa: &core.DataSource{
@@ -232,7 +246,36 @@
 		}
 	}

-	return secret
+	return secret, nil
+}
+
+func buildSPIFFECertValidatorConfig(s *security.SecretItem) (*tls.Secret_ValidationContext, error) {
+	var configTrustDomains []*tls.SPIFFECertValidatorConfig_TrustDomain
+
+	for _, bundle := range s.TrustBundles.Bundles() {
+		caBytes, err := bundle.Marshal()
+		if err != nil {
+			return nil, err
+		}
+		configTrustDomains = append(configTrustDomains, &tls.SPIFFECertValidatorConfig_TrustDomain{
+			Name: bundle.TrustDomain().String(),
+			TrustBundle: &core.DataSource{Specifier: &core.DataSource_InlineBytes{
+				InlineBytes: caBytes,
+			}},
+		})
+	}
+
+	typedConfig, err := anypb.New(&tls.SPIFFECertValidatorConfig{TrustDomains: configTrustDomains})
+	if err != nil {
+		return nil, err
+	}
+
+	return &tls.Secret_ValidationContext{ValidationContext: &tls.CertificateValidationContext{
+		CustomValidatorConfig: &core.TypedExtensionConfig{
+			Name:        "envoy.tls.cert_validator.spiffe",
+			TypedConfig: typedConfig,
+		},
+	}}, nil
 }

 func pushLog(names []string) model.XdsLogDetails {
diff --git a/pilot/pkg/bootstrap/certcontroller.go b/pilot/pkg/bootstrap/certcontroller.go
--- a/pilot/pkg/bootstrap/certcontroller.go	(revision 8f9263961b52f11f9bfb731e0eeae096bb0acfd0)
+++ b/pilot/pkg/bootstrap/certcontroller.go	(date 1636568293868)
@@ -290,3 +290,32 @@
 	s.certMu.Unlock()
 	return nil
 }
+
+func (s *Server) setIstioCertBundleAndNotify(certChain []byte, key []byte, bundle []byte) {
+	keyPair, err := tls.X509KeyPair(certChain, key)
+
+	if err != nil {
+		log.Errorf("istiod loading x509 key pairs failed: %v", err)
+		return
+	}
+	for _, c := range keyPair.Certificate {
+		x509Cert, err := x509.ParseCertificates(c)
+		if err != nil {
+			log.Errorf("x509 cert - ParseCertificates() error: %v", err)
+			return
+		}
+		for _, c := range x509Cert {
+			log.Infof("x509 cert - Issuer: %q, Subject: %q, SN: %x, NotBefore: %q, NotAfter: %q",
+				c.Issuer, c.Subject, c.SerialNumber,
+				c.NotBefore.Format(time.RFC3339), c.NotAfter.Format(time.RFC3339))
+		}
+	}
+
+	s.certMu.Lock()
+	s.istiodCert = &keyPair
+	s.certMu.Unlock()
+	log.Info("istiod certificates are set")
+
+	s.istiodCertBundleWatcher.SetAndNotify(nil, nil, bundle)
+	log.Info("istiod Cert Bundle Watcher notified")
+}
diff --git a/security/pkg/nodeagent/cache/spiffesecret.go b/security/pkg/nodeagent/cache/spiffesecret.go
new file mode 100644
--- /dev/null	(date 1636568293868)
+++ b/security/pkg/nodeagent/cache/spiffesecret.go	(date 1636568293868)
@@ -0,0 +1,217 @@
+package cache
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/go-spiffe/v2/workloadapi"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"istio.io/istio/pkg/security"
+	"istio.io/pkg/log"
+)
+
+// SpiffeSecretManager is a source of SecretItems (X.509 SVIDs and trust bundles) maintained via the
+// Workload API. Implements the
+type SpiffeSecretManager struct {
+	sync.RWMutex
+	trustDomain       spiffeid.TrustDomain
+	configTrustBundle []byte
+	secretItem        *security.SecretItem
+	notifyCallback    func(resourceName string)
+	cancelWatcher     context.CancelFunc
+	updatedCh         chan struct{}
+}
+
+// NewSpiffeSecretManager creates a new SpiffeSecretManager. It blocks until the initial update
+// has been received from the Workload API.
+func NewSpiffeSecretManager(opt *security.Options) (*SpiffeSecretManager, error) {
+	td, err := spiffeid.TrustDomainFromString(opt.TrustDomain)
+	if err != nil {
+		return nil, fmt.Errorf("error trying to parse trust domain %q reason: %v", opt.TrustDomain, err)
+	}
+
+	sm := &SpiffeSecretManager{
+		trustDomain: td,
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	sm.cancelWatcher = cancel
+	sm.updatedCh = make(chan struct{})
+
+	go sm.watcherTask(ctx)
+
+	err = sm.WaitUntilUpdated(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing the SPIFFE secret manager")
+	}
+
+	return sm, nil
+}
+
+// WaitUntilUpdated waits until the secret manager is updated or the context is done,
+// in which case ctx.Err() is returned.
+func (w *SpiffeSecretManager) WaitUntilUpdated(ctx context.Context) error {
+	select {
+	case <-w.updatedCh:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Updated returns a channel that is sent on whenever the secret manager is updated.
+func (w *SpiffeSecretManager) Updated() <-chan struct{} {
+	return w.updatedCh
+}
+
+// GenerateSecret generates a SecretItem for the given resourceName (default or ROOTCA).
+func (s *SpiffeSecretManager) GenerateSecret(resourceName string) (*security.SecretItem, error) {
+	s.RLock()
+	defer s.RUnlock()
+
+	si := s.secretItem
+	if si == nil {
+		return nil, fmt.Errorf("secret was not in cache for resource: %v", resourceName)
+	}
+
+	if resourceName == security.RootCertReqResourceName {
+
+		ns := &security.SecretItem{
+			ResourceName: resourceName,
+			RootCert:     si.RootCert,
+			// adding all trust bundles
+			TrustBundles: si.TrustBundles,
+		}
+
+		cacheLog.WithLabels("ttl", time.Until(si.ExpireTime)).Info("returned workload trust anchor from cache")
+		return ns, nil
+	}
+
+	ns := &security.SecretItem{
+		ResourceName:     resourceName,
+		CertificateChain: si.CertificateChain,
+		PrivateKey:       si.PrivateKey,
+		ExpireTime:       si.ExpireTime,
+		CreatedTime:      si.CreatedTime,
+	}
+	cacheLog.WithLabels("ttl", time.Until(si.ExpireTime)).Info("returned workload certificate from cache")
+	return ns, nil
+}
+
+// UpdateConfigTrustBundle updates the configTrustBundle and calls the notify callback function.
+func (s *SpiffeSecretManager) UpdateConfigTrustBundle(trustBundle []byte) error {
+	log.WithLabels("UpdateConfigTrustBundle").Info(string(trustBundle))
+	s.Lock()
+	defer s.Unlock()
+
+	if bytes.Equal(s.configTrustBundle, trustBundle) {
+		return nil
+	}
+	s.configTrustBundle = trustBundle
+	s.callUpdateCallback(security.RootCertReqResourceName)
+	return nil
+}
+
+// Close closes the SPIFFE secret manager instance.
+func (s *SpiffeSecretManager) Close() {
+	if s.cancelWatcher != nil {
+		log.Info("closing SPIFFE secret manager")
+		s.cancelWatcher()
+	}
+}
+
+// SetUpdateCallback configures the manager with a notify callback function.
+func (s *SpiffeSecretManager) SetUpdateCallback(f func(resourceName string)) {
+	s.Lock()
+	defer s.Unlock()
+	s.notifyCallback = f
+}
+
+// OnX509ContextUpdate is run every time a new update is pushed by the SPIFFE Workload API.
+func (s *SpiffeSecretManager) OnX509ContextUpdate(c *workloadapi.X509Context) {
+	log.Info("got new identities from the SPIFFE Workload API")
+	if len(c.SVIDs) < 1 {
+		log.Error("identities were not found on workload API response")
+		return
+	}
+	if len(c.SVIDs[0].Certificates) < 1 {
+		log.Error("leaf certificate was not found on workload API response")
+		return
+	}
+
+	svid := c.DefaultSVID()
+	workloadChain, workloadKey, err := svid.Marshal()
+	if err != nil {
+		log.Fatalf("unable to marshal X.509 SVID: %v", err)
+		return
+	}
+
+	bundle, ok := c.Bundles.Get(s.trustDomain)
+	if !ok {
+		log.WithLabels("trust_domain", s.trustDomain).Fatal("unable to get trust bundle for trust domain")
+		return
+	}
+
+	root, err := bundle.Marshal()
+	if err != nil {
+		log.Fatalf("unable to marshal trust bundle: %v", err)
+		return
+	}
+
+	certChain := concatCerts([]string{string(workloadChain)})
+	leaf := c.SVIDs[0].Certificates[0]
+
+	item := &security.SecretItem{
+		CertificateChain: certChain,
+		PrivateKey:       workloadKey,
+		RootCert:         root,
+		TrustBundles:     c.Bundles,
+		ResourceName:     security.WorkloadKeyCertResourceName,
+		CreatedTime:      leaf.NotBefore,
+		ExpireTime:       leaf.NotAfter,
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	if s.secretItem == nil || !bytes.Equal(s.secretItem.RootCert, item.RootCert) {
+		s.callUpdateCallback(security.RootCertReqResourceName)
+	}
+	if s.secretItem == nil || !bytes.Equal(s.secretItem.CertificateChain, item.CertificateChain) {
+		s.callUpdateCallback(security.WorkloadKeyCertResourceName)
+	}
+	s.secretItem = item
+
+	select {
+	case s.updatedCh <- struct{}{}:
+		log.Info("notify message sent on updateCh")
+	default:
+		log.Info("notify message dropped")
+	}
+}
+
+// OnX509ContextWatchError is run when the client runs into an error.
+func (s *SpiffeSecretManager) OnX509ContextWatchError(err error) {
+	if status.Code(err) != codes.Canceled {
+		log.Infof("error calling SPIFE Workload API: %v", err)
+	}
+}
+
+func (s *SpiffeSecretManager) callUpdateCallback(resourceName string) {
+	log.WithLabels("resource", resourceName).Info("fetched new identity from SPIFFE Workload API")
+	if s.notifyCallback != nil {
+		s.notifyCallback(resourceName)
+	}
+}
+
+func (s *SpiffeSecretManager) watcherTask(ctx context.Context) {
+	err := workloadapi.WatchX509Context(ctx, s)
+	if err != nil && status.Code(err) != codes.Canceled {
+		log.Fatalf("error watching SPIFFE workload API: %v", err)
+	}
+}
